File: design/HACC/hacc/payload_builder.py
from __future__ import annotations

from .ir import AguConfig, BlockDesc, CompilerContext, DmaRule, JobDesc, NluRule, PatchEntry, ProfileConfig

PROFILE_WORDS = 3 + 3 * 7
AGU_WORDS = 7
DMA_RULE_WORDS = 5
NLU_RULE_WORDS = 5
BLOCK_WORDS = 19
PATCH_WORDS = 5
JOB_WORDS = 19


class PayloadBuilder:
    @staticmethod
    def build(ctx: CompilerContext) -> None:
        package = ctx.package
        package.profile_section = PayloadBuilder.serialize_profiles(ctx.profiles)
        package.agu_section = PayloadBuilder.serialize_agus(ctx.agus)
        package.dma_section = PayloadBuilder.serialize_dma_rules(ctx.dma_rules)
        package.nlu_section = PayloadBuilder.serialize_nlu_rules(ctx.nlu_rules)
        package.block_section = PayloadBuilder.serialize_blocks(ctx.blocks)
        package.patch_section = PayloadBuilder.serialize_patches(ctx.patches)
        package.pe_section = list(ctx.pe_payload.words)
        package.scan_section = list(ctx.scan_payload.words)

        job = JobDesc()
        cursor = JOB_WORDS * 4

        job.block_table_base = cursor
        job.block_count = len(ctx.blocks)
        cursor += len(package.block_section) * 4

        job.profile_table_base = cursor
        job.profile_count = len(ctx.profiles)
        cursor += len(package.profile_section) * 4

        job.dma_table_base = cursor
        job.dma_count = len(ctx.dma_rules)
        cursor += len(package.dma_section) * 4

        job.agu_table_base = cursor
        job.agu_count = len(ctx.agus)
        required_cluster_mask = 0
        for block in ctx.blocks:
            required_cluster_mask |= block.cluster_mask
        job.required_cluster_mask = required_cluster_mask or ctx.schedule.cluster_mask

        job.nlu_table_base = cursor
        job.nlu_count = len(ctx.nlu_rules)
        cursor += len(package.nlu_section) * 4

        job.pe_table_base = cursor
        job.pe_count = len(package.pe_section)
        cursor += len(package.pe_section) * 4

        job.scan_table_base = cursor
        job.scan_count = len(package.scan_section)
        cursor += len(package.scan_section) * 4

        job.patch_table_base = cursor
        job.patch_count = len(ctx.patches)
        job.required_cluster_mask = ctx.schedule.cluster_mask
        job.required_caps = 0x3F
        package.job_section = PayloadBuilder.serialize_job(job)

    @staticmethod
    def serialize_job(job: JobDesc) -> list[int]:
        return [
            job.block_table_base,
            job.block_count,
            job.profile_table_base,
            job.profile_count,
            job.dma_table_base,
            job.dma_count,
            job.agu_table_base,
            job.agu_count,
            job.nlu_table_base,
            job.nlu_count,
            job.pe_table_base,
            job.pe_count,
            job.scan_table_base,
            job.scan_count,
            job.patch_table_base,
            job.patch_count,
            job.required_cluster_mask,
            job.required_caps,
            job.job_flags,
        ]

    @staticmethod
    def serialize_blocks(blocks: list[BlockDesc]) -> list[int]:
        output: list[int] = []
        for block in blocks:
            output.append(block.loop_rank)
            output.extend(block.loop_extent)
            output.extend(
                [
                    block.repeat_count,
                    block.cluster_mask,
                    block.profile_rule_idx,
                    block.dma_rule_idx,
                    block.agu_rule_idx,
                    block.pe_payload_idx,
                    block.scan_payload_idx,
                    block.nlu_rule_idx,
                    block.rule_stride,
                    block.nlu_rule_stride,
                    block.total_waves,
                    block.patch_begin,
                    block.patch_count,
                    block.block_flags,
                ]
            )
        return output

    @staticmethod
    def serialize_profiles(profiles: list[ProfileConfig]) -> list[int]:
        output: list[int] = []
        for profile in profiles:
            output.extend([profile.pe_mode, profile.pe_rows, profile.pe_cols])
            for agu in (profile.agu_ifmap, profile.agu_weight, profile.agu_ofmap):
                output.extend(PayloadBuilder.serialize_single_agu(agu))
        return output

    @staticmethod
    def serialize_single_agu(agu: AguConfig) -> list[int]:
        return [agu.base_addr, agu.iter0, agu.stride0, agu.iter1, agu.stride1, agu.iter2, agu.stride2]

    @staticmethod
    def serialize_agus(agus: list[AguConfig]) -> list[int]:
        output: list[int] = []
        for agu in agus:
            output.extend(PayloadBuilder.serialize_single_agu(agu))
        return output

    @staticmethod
    def serialize_dma_rules(rules: list[DmaRule]) -> list[int]:
        output: list[int] = []
        for rule in rules:
            output.extend([rule.mode, rule.cluster_mask, rule.word_count, rule.src_addr, rule.dst_addr])
        return output

    @staticmethod
    def serialize_nlu_rules(rules: list[NluRule]) -> list[int]:
        output: list[int] = []
        for rule in rules:
            output.extend([rule.mode, rule.src_addr, rule.dst_addr, rule.word_count, rule.nlu_id])
        return output

    @staticmethod
    def serialize_patches(patches: list[PatchEntry]) -> list[int]:
        output: list[int] = []
        for patch in patches:
            output.extend([patch.wave_id, patch.valid_mask, patch.profile_idx, patch.dma_idx, patch.cluster_mask])
        return output

File: design/HACC/src/payload_builder.cpp
/**
 * @file payload_builder.cpp
 * @brief HACC Compiler Runtime Payload Builder implementation.
 */
#include "hacc/payload_builder.hpp"

namespace hacc {

/**
 * @brief Number of 32-bit words per serialized ProfileConfig.
 *
 * Layout: pe_mode, pe_rows, pe_cols, then 3 AGU configs (7 words each) = 24 words.
 */
static constexpr uint32_t PROFILE_WORDS = 3 + 3 * 7;

/** @brief Number of 32-bit words per serialized AguConfig. */
static constexpr uint32_t AGU_WORDS = 7;

/** @brief Number of 32-bit words per serialized DmaRule. */
static constexpr uint32_t DMA_RULE_WORDS = 5;

/** @brief Number of 32-bit words per serialized NluRule. */
static constexpr uint32_t NLU_RULE_WORDS = 5;

/** @brief Number of 32-bit words per serialized BlockDesc. */
static constexpr uint32_t BLOCK_WORDS = 19;

/** @brief Number of 32-bit words per serialized PatchEntry. */
static constexpr uint32_t PATCH_WORDS = 5;

/** @brief Number of 32-bit words per serialized JobDesc. */
static constexpr uint32_t JOB_WORDS = 19;

void PayloadBuilder::build(CompilerContext &ctx) {
    auto &pkg = ctx.package;

    /* Serialize all tables. */
    pkg.profile_section = serialize_profiles(ctx.profiles);
    pkg.agu_section     = serialize_agus(ctx.agus);
    pkg.dma_section     = serialize_dma_rules(ctx.dma_rules);
    pkg.nlu_section     = serialize_nlu_rules(ctx.nlu_rules);
    pkg.block_section   = serialize_blocks(ctx.blocks);
    pkg.patch_section   = serialize_patches(ctx.patches);
    pkg.pe_section      = ctx.pe_payload.words;
    pkg.scan_section    = ctx.scan_payload.words;

    /* Build job descriptor with table layout offsets. */
    JobDesc job;
    uint32_t cursor = JOB_WORDS * 4; /* Start after job section itself. */

    job.block_table_base   = cursor;
    job.block_count        = static_cast<uint32_t>(ctx.blocks.size());
    cursor += static_cast<uint32_t>(pkg.block_section.size()) * 4;

    job.profile_table_base = cursor;
    job.profile_count      = static_cast<uint32_t>(ctx.profiles.size());
    cursor += static_cast<uint32_t>(pkg.profile_section.size()) * 4;

    job.dma_table_base     = cursor;
    job.dma_count          = static_cast<uint32_t>(ctx.dma_rules.size());
    cursor += static_cast<uint32_t>(pkg.dma_section.size()) * 4;

    job.agu_table_base     = cursor;
    job.agu_count          = static_cast<uint32_t>(ctx.agus.size());
    cursor += static_cast<uint32_t>(pkg.agu_section.size()) * 4;

    job.nlu_table_base     = cursor;
    job.nlu_count          = static_cast<uint32_t>(ctx.nlu_rules.size());
    cursor += static_cast<uint32_t>(pkg.nlu_section.size()) * 4;

    job.pe_table_base      = cursor;
    job.pe_count           = static_cast<uint32_t>(pkg.pe_section.size());
    cursor += static_cast<uint32_t>(pkg.pe_section.size()) * 4;

    job.scan_table_base    = cursor;
    job.scan_count         = static_cast<uint32_t>(pkg.scan_section.size());
    cursor += static_cast<uint32_t>(pkg.scan_section.size()) * 4;

    job.patch_table_base   = cursor;
    job.patch_count        = static_cast<uint32_t>(ctx.patches.size());

    job.required_cluster_mask = ctx.schedule.cluster_mask;
    job.required_caps      = 0x3F; /* All capabilities required */
    job.job_flags          = 0;

    pkg.job_section = serialize_job(job);
}

std::vector<uint32_t> PayloadBuilder::serialize_job(const JobDesc &job) {
    return {
        job.block_table_base,  job.block_count,
        job.profile_table_base, job.profile_count,
        job.dma_table_base,    job.dma_count,
        job.agu_table_base,    job.agu_count,
        job.nlu_table_base,    job.nlu_count,
        job.pe_table_base,     job.pe_count,
        job.scan_table_base,   job.scan_count,
        job.patch_table_base,  job.patch_count,
        job.required_cluster_mask,
        job.required_caps,
        job.job_flags,
    };
}

std::vector<uint32_t> PayloadBuilder::serialize_blocks(const std::vector<BlockDesc> &blocks) {
    std::vector<uint32_t> out;
    out.reserve(blocks.size() * BLOCK_WORDS);
    for (const auto &b : blocks) {
        out.push_back(b.loop_rank);
        for (auto e : b.loop_extent) out.push_back(e);
        out.push_back(b.repeat_count);
        out.push_back(b.cluster_mask);
        out.push_back(b.profile_rule_idx);
        out.push_back(b.dma_rule_idx);
        out.push_back(b.agu_rule_idx);
        out.push_back(b.pe_payload_idx);
        out.push_back(b.scan_payload_idx);
        out.push_back(b.nlu_rule_idx);
        out.push_back(b.rule_stride);
        out.push_back(b.nlu_rule_stride);
        out.push_back(b.total_waves);
        out.push_back(b.patch_begin);
        out.push_back(b.patch_count);
        out.push_back(b.block_flags);
    }
    return out;
}

std::vector<uint32_t> PayloadBuilder::serialize_profiles(
        const std::vector<ProfileConfig> &profiles) {
    std::vector<uint32_t> out;
    out.reserve(profiles.size() * PROFILE_WORDS);
    for (const auto &p : profiles) {
        out.push_back(p.pe_mode);
        out.push_back(p.pe_rows);
        out.push_back(p.pe_cols);

        auto push_agu = [&](const AguConfig &a) {
            out.push_back(a.base_addr);
            out.push_back(a.iter0);
            out.push_back(a.stride0);
            out.push_back(a.iter1);
            out.push_back(a.stride1);
            out.push_back(a.iter2);
            out.push_back(a.stride2);
        };
        push_agu(p.agu_ifmap);
        push_agu(p.agu_weight);
        push_agu(p.agu_ofmap);
    }
    return out;
}

std::vector<uint32_t> PayloadBuilder::serialize_agus(const std::vector<AguConfig> &agus) {
    std::vector<uint32_t> out;
    out.reserve(agus.size() * AGU_WORDS);
    for (const auto &a : agus) {
        out.push_back(a.base_addr);
        out.push_back(a.iter0);
        out.push_back(a.stride0);
        out.push_back(a.iter1);
        out.push_back(a.stride1);
        out.push_back(a.iter2);
        out.push_back(a.stride2);
    }
    return out;
}

std::vector<uint32_t> PayloadBuilder::serialize_dma_rules(const std::vector<DmaRule> &rules) {
    std::vector<uint32_t> out;
    out.reserve(rules.size() * DMA_RULE_WORDS);
    for (const auto &r : rules) {
        out.push_back(r.mode);
        out.push_back(r.cluster_mask);
        out.push_back(r.word_count);
        out.push_back(r.src_addr);
        out.push_back(r.dst_addr);
    }
    return out;
}

std::vector<uint32_t> PayloadBuilder::serialize_nlu_rules(const std::vector<NluRule> &rules) {
    std::vector<uint32_t> out;
    out.reserve(rules.size() * NLU_RULE_WORDS);
    for (const auto &r : rules) {
        out.push_back(r.mode);
        out.push_back(r.src_addr);
        out.push_back(r.dst_addr);
        out.push_back(r.word_count);
        out.push_back(r.nlu_id);
    }
    return out;
}

std::vector<uint32_t> PayloadBuilder::serialize_patches(const std::vector<PatchEntry> &patches) {
    std::vector<uint32_t> out;
    out.reserve(patches.size() * PATCH_WORDS);
    for (const auto &p : patches) {
        out.push_back(p.wave_id);
        out.push_back(p.valid_mask);
        out.push_back(p.profile_idx);
        out.push_back(p.dma_idx);
        out.push_back(p.cluster_mask);
    }
    return out;
}

} // namespace hacc
